data-otservbr-global/monster/quests/primal_ordeal_quest/the_primal_menace.lua

```diff
@@ -81,24 +130,217 @@ monster.defenses = {
+		local spawnPosition = getSpawnPosition(monster)
+		spawnTimer(monster, spawnPosition, spawnPod)
+	end
+end
+
+local function handlePodSpawn(monster, hazardPoints)
+	local nextSpawn = monster:getStorageValue(thePrimalMenaceConfig.Storage.NextPodSpawn)
+	if (nextSpawn - os.time() < 0) then
+		spawnPods(monster, hazardPoints)
+
+		setNextTimeToSpawn(monster, thePrimalMenaceConfig.Storage.NextPodSpawn, thePrimalMenaceConfig.PodConfig, hazardPoints)
+	end
+end
+
+local function spawnMonster(monster, spawnPosition)
+	local randomMonsterIndex = math.random(#thePrimalMenaceConfig.MonsterConfig.MonsterPool)
+	local primalBeastEntry = {
+		Monster = Game.createMonster(thePrimalMenaceConfig.MonsterConfig.MonsterPool[randomMonsterIndex], spawnPosition),
+		Created = os.time()
+	}
+	local monsterMaxHealth = primalBeastEntry.Monster:getMaxHealth()
+	primalBeastEntry.Monster:setHealth(monsterMaxHealth * thePrimalMenaceConfig.MonsterConfig.HpRateOnSpawn)
+
+	local primalBeasts = monster:getStorageValue(thePrimalMenaceConfig.Storage.PrimalBeasts)
+	table.insert(primalBeasts, primalBeastEntry)
+	monster:setStorageValue(thePrimalMenaceConfig.Storage.PrimalBeasts, primalBeasts)
+end
+
+local function spawnMonsters(monster, hazardPoints)
+	local count = spawnCount(thePrimalMenaceConfig.MonsterConfig, hazardPoints)
+	for i = 1, count do
+		local spawnPosition = getSpawnPosition(monster)
+		spawnTimer(monster, spawnPosition, spawnMonster)
+	end
+end
+
+local function handleMonsterSpawn(monster, hazardPoints)
+	local nextSpawn = monster:getStorageValue(thePrimalMenaceConfig.Storage.NextMonsterSpawn)
+	if (nextSpawn - os.time() < 0) then
+		spawnMonsters(monster, hazardPoints)
+
+		setNextTimeToSpawn(monster, thePrimalMenaceConfig.Storage.NextMonsterSpawn, thePrimalMenaceConfig.MonsterConfig, hazardPoints)
+	end
+end
+
+local function handlePrimalBeasts(monster)
+	primalBeasts = monster:getStorageValue(thePrimalMenaceConfig.Storage.PrimalBeasts)
+	indexesToRemove = {}
+
+	for index, beastData in pairs(primalBeasts) do
+		local monster = beastData.Monster
+		local created = beastData.Created
+		if not monster:getHealth() or monster:getHealth() == 0 then
+			table.insert(indexesToRemove, index)
+		elseif (os.time() - created > 20 and monster:getHealth() > 0) then
+			local position = monster:getPosition()
+			monster:remove()
+			table.insert(indexesToRemove, index)
+			if position then
+				Game.createMonster("Fungosaurus", position)
+			end
+		end
+	end
+
+	for i = #indexesToRemove, 1, -1 do
+		local indexToRemove = indexesToRemove[i]
+		table.remove(primalBeasts, indexToRemove)
+	end
+
+	monster:setStorageValue(thePrimalMenaceConfig.Storage.PrimalBeasts, primalBeasts)
+end
+
+mType.onThink = function(monster, interval)
+	if (monster:getStorageValue(thePrimalMenaceConfig.Storage.Initialized) == -1) then
+		initialize(monster)
+	end
+
+	local hazardPoints = getHazardPoints(monster)
+	handleMonsterSpawn(monster, hazardPoints)
+	handlePodSpawn(monster, hazardPoints)
```

Comment:
Be careful when passing pointers to addEvent, it is potentially harmful behavior and should crash.

Reply:
Added ids when calling the addEvent instead. Hope this is what you meant.